pkg/storage/sqlcommon/sqlcommon.go

```diff
@@ -187,11 +190,35 @@ func NewSQLTupleIterator(rows *sql.Rows) *SQLTupleIterator {
 		rows:     rows,
 		resultCh: make(chan *storage.TupleRecord, 1),
 		errCh:    make(chan error, 1),
+		firstRow: nil, // GUARDED_BY(mu). The firstRow is used as a temporary storage place if head is called.
+		// If firstRow is nil and Head is called, rows.Next() will return the first item and advance
+		// the iterator. Thus, we will need to store this first item so that future Head() and Next()
+		// will use this item instead. Otherwise, the first item will be lost.
+		mu: sync.Mutex{},
 	}
 }
 
 func (t *SQLTupleIterator) next() (*storage.TupleRecord, error) {
+	t.mu.Lock()
```

Comment:
I'm not seeing a good reason to have `t.mu.Unlock` in 3 lines, instead of having `defer t.mu.Unlock` right below this line. It prevents bugs where you forget to unlock the mutex.

Reply:
want to minimize the time mu lock is being hold (especially for proto.Unmarshal in the condition as I am not sure how long that will take).

Reply:
I see. What about something like this?

```go
func (t *SQLTupleIterator) next() (*storage.TupleRecord, error) {
	t.mu.RLock() // lock for reads
	firstRowAvaible := t.firstRow != nil
	t.mu.RUnlock()

	if firstRowAvaible {
		firstRow := t.firstRow
		t.mu.Lock() // lock for writes
		t.firstRow = nil
		t.mu.Unlock()
		return firstRow, nil
	}

```

Reply:
t.firstRow may potentially be updated in between.